File: pydsfapi/pydsfapi/commands/code.py
"""
Module code contains relevant classes and enums being received
as well as sent back to the server.

    Python interface to DuetSoftwareFramework
    Copyright (C) 2020 Duet3D

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
from enum import Enum, IntEnum
from .basecommands import BaseCommand
from .codeparameter import CodeParameter
from .result import Message


class CodeType(str, Enum):
    """Type of a generic G/M/T-code. If none is applicable, it is treated as a comment"""
    Comment = 'Q'
    GCode = 'G'
    MCode = 'M'
    TCode = 'T'


class KeywordType(IntEnum):
    """Enumeration of conditional G-code keywords"""
    KeywordNone = 0
    If = 1
    ElseIf = 2
    Else = 3
    While = 4
    Break = 5
    Return = 6
    Abort = 7
    Var = 8
    Set = 9
    Echo = 10
    Continue = 11


keyword_type_names = {
        KeywordType.Abort: 'abort',
        KeywordType.Break: 'break',
        KeywordType.Echo: 'echo',
        KeywordType.Else: 'else',
        KeywordType.ElseIf: 'elif',
        KeywordType.If: 'if',
        KeywordType.Return: 'return',
        KeywordType.Set: 'set',
        KeywordType.Var: 'var',
        KeywordType.While: 'while',
        KeywordType.Continue: 'continue',
}


class CodeFlags(IntEnum):
    """Code bits to classify G/M/T-codes"""
    CodeFlagsNone = 0
    Asynchronous = 1
    IsPreProcessed = 2
    IsPostProcessed = 4
    IsFromMacro = 8
    IsNestedMacro = 16
    IsFromConfig = 32
    IsFromConfigOverride = 64
    EnforceAbsolutePosition = 128
    IsPrioritized = 256
    Unbuffered = 512
    IsFromFirmware = 1024


class Code(BaseCommand):
    """A parsed representation of a generic G/M/T-code"""
    @classmethod
    def from_json(cls, data):
        """Deserialize an instance of this class from JSON deserialized dictionary"""
        data['result'] = [] if data['result'] is None else list(
            map(Message.from_json, data['result']))
        data['parameters'] = list(map(CodeParameter.from_json, data['parameters']))
        return cls(**data)

    def parameter(self, letter: str, default=None):
        """Retrieve the parameter whose letter equals c or generate a default parameter"""
        letter = letter.upper()
        param = [param for param in self.parameters if param.letter.upper() == letter]
        if len(param) > 0:
            return param[0]
        if default is not None:
            return CodeParameter.simple_param(letter, default)
        return None

    def get_unprecedented_string(self, quote: bool = False):
        """
        Reconstruct an unprecedented string from the parameter list or
        retrieve the parameter which does not have a letter assigned.
        """
        str_list = []
        for param in self.parameters:
            if quote and param.is_string:
                str_list.append('{0}"{1}"'.format(param.letter, param.string_value))
            else:
                str_list.append('{0}{1}'.format(param.letter, param.string_value))
        return ' '.join(str_list)

    def __str__(self):
        """Convert the parsed code back to a text-based G/M/T-code"""
        if self.keyword != KeywordType.KeywordNone:
            if self.keywordArgument is not None:
                return '{0} {1}'.format(self.keyword_to_str(), self.keywordArgument)
            else:
                return self.keyword_to_str()

        if self.type == CodeType.Comment:
            return ';{0}'.format(self.comment)

        str_list = []
        str_list.append(self.short_str())

        for param in self.parameters:
            str_list.append(' {0}'.format(param))

        if self.comment:
            if len(str_list) > 0:
                str_list.append(' ')
            str_list.append(';{0}'.format(self.comment))

        if len(self.result) > 0:
            str_list.append(' => {0}'.format(self.result))

        return ''.join(str_list)

    def short_str(self):
        """Convert only the command portion to a text-based G/M/T-code (e.g. G28)"""
        if self.type == CodeType.Comment:
            return '(comment)'

        prefix = 'G53 ' if self.flags & CodeFlags.EnforceAbsolutePosition != 0 else ''
        if self.majorNumber is not None:
            if self.minorNumber is not None:
                return '{0}{1}{2}.{3}'.format(prefix, self.type, self.majorNumber, self.minorNumber)

            return '{0}{1}{2}'.format(prefix, self.type, self.majorNumber)

        return '{0}{1}'.format(prefix, self.type)

    def keyword_to_str(self):
        """Convert the keyword to a str"""
        return keyword_type_names.get(self.keyword)

File: godsfapi/v3/commands/code.go
// Deprecated: This package was deprected, please visit https://github.com/Duet3D/dsf-go.
package commands

import (
	"fmt"
	"strings"

	"github.com/Duet3D/DSF-APIs/godsfapi/v3/machine/messages"
	"github.com/Duet3D/DSF-APIs/godsfapi/v3/types"
)

// CodeResult is a list of code results
// Deprecated: This class is now deprecated. It will be replaced with messages.Message in foreseeable future
type CodeResult []messages.Message

func (cr CodeResult) String() string {
	var b strings.Builder
	for _, m := range cr {
		if m.Content != "" {
			b.WriteString(m.String())
			b.WriteByte('\n')
		}
	}
	return b.String()
}

// CodeFlags are bit masks to classify G/M/T-codes
type CodeFlags int64

const (
	// Asynchronous codes are considered finished as soon as they enter the code queue
	Asynchronous CodeFlags = 1 << iota
	// IsPreProcessed marks pre-processed codes
	IsPreProcessed
	// IsPostProcessed marks post-processed codes
	IsPostProcessed
	// IsFromMacro indicates code originating from macro
	IsFromMacro
	// IsNestedMacro indicates code originating from system macro
	IsNestedMacro
	// IsFromConfig indicates code originating from config.g or config.g.bak
	IsFromConfig
	// IsFromConfigOverride indicated code originating from config-override.g
	IsFromConfigOverride
	// EnforceAbsolutePosition marks code prefixed with G53
	EnforceAbsolutePosition
	// IsPrioritized will be sent to the firmware as soon as possible jumping all queued codes
	IsPrioritized
	// Unbuffered will execute this code circumventing any buffers
	// Do NOT process another code on the same channel before this code has been fully executed
	Unbuffered
	// IsFromFirmware indicates if this code was requested by the firmware
	IsFromFirmware
	// IsLastCode indicates if this is the last code on the line
	IsLastCode
	// CodeFlagsNone is a placeholder to indicate that no flags are set
	CodeFlagsNone = 0
)

// CodeType is the generic type of G/M/T-code or being a comment
type CodeType string

const (
	// Comment if this code is whole line comment
	Comment CodeType = "Q"
	// GCode if this code is a G-Code
	GCode = "G"
	// MCode if this code is a M-Code
	MCode = "M"
	// TCode if this code is a T-Code
	TCode = "T"
)

// KeywordType is the type of conditional G-code
type KeywordType byte

const (
	// None for no conditional code
	None KeywordType = iota
	// If condition
	If
	// ElseIf condition
	ElseIf
	// Else condition
	Else
	// While condition
	While
	// Break instruction (used in While)
	Break
	// Return instruction
	// Deprecated: was never supported by RRF
	Return
	// Abort instruction
	Abort
	// Var operation
	Var
	// Set operation
	Set
	// Echo operation
	Echo
	// Continue instruction (used in While)
	Continue
	// Global operation
	Global
)

func (k KeywordType) String() string {
	switch k {
	case Abort:
		return "abort"
	case Break:
		return "break"
	case Echo:
		return "echo"
	case Else:
		return "else"
	case ElseIf:
		return "elif"
	case If:
		return "if"
	case Return:
		return "return"
	case Set:
		return "set"
	case Var:
		return "var"
	case While:
		return "while"
	default:
		return ""
	}
}

// Code is a parsed representation of a generic G/M/T/code
type Code struct {
	BaseCommand
	// SourceConnection ID this code was received from. If this is 0, the code originates from an internal DCS task
	// Usually there is no need to populate this property. It is internally overwritten by the control server on receipt
	SourceConnection int64
	// Result of this code. This property is only set when the code has finished its execution
	// It remains nil if the code has been cancelled.
	Result CodeResult
	// Type of the code
	Type CodeType
	// Channel to send this code to
	Channel types.CodeChannel
	// LineNumber of this code
	LineNumber *int64
	// Indent are the number of whitespaces prefixing the command content
	Indent byte
	// Keyword type of conditional G-code
	Keyword KeywordType
	// KeywordArgument of the conditional G-code
	KeywordArgument string
	// MajorNumber of the code (e.g. 28 in G28)
	MajorNumber *int64
	// MinorNumber of the code (e.g. 3 in G54.3)
	MinorNumber *int8
	// Flags of this code
	Flags CodeFlags
	// Comment provided to this G/M/T-code
	Comment string
	// FilePosition of this code in bytes (optional)
	FilePosition *int64
	// Length of the original code in bytes (optional)
	Length *int64
	// Parameters are a list of parsed code parameters
	Parameters []CodeParameter
}

// NewCode instantiates a Code with default values
func NewCode() *Code {
	return &Code{
		BaseCommand: *NewBaseCommand("Code"),
		Type:        Comment,
		Channel:     types.DefaultChannel,
		Keyword:     None,
		Flags:       CodeFlagsNone,
	}
}

// Clone an existing Code into a new instance
func (c *Code) Clone() *Code {
	cc := *c
	cparams := make([]CodeParameter, 0)
	for _, p := range c.Parameters {
		cp := p.Clone()
		cparams = append(cparams, *cp)
	}
	cc.Parameters = cparams
	return &cc
}

// IsMajorNumber is a convenience function that checks if
// the MajorNumber of this Code instance is present and equal
// to the given value
func (c *Code) IsMajorNumber(n int64) bool {
	return c.MajorNumber != nil && *c.MajorNumber == n
}

// HasFlag checks if this code has the given flag set
func (c *Code) HasFlag(flag CodeFlags) bool {
	return (c.Flags & flag) != 0
}

// HasParameter returns whether or not a certain parameter is present without returning the
// CodeParameter instance
func (c *Code) HasParameter(letter string) bool {
	return c.Parameter(letter) != nil
}

// Parameter retrieves a parameter for the given letter. This will return nil in case there
// is no parameter with this letter. Lookup is case-insensitive.
func (c *Code) Parameter(letter string) *CodeParameter {
	l := strings.ToUpper(letter)
	for _, p := range c.Parameters {
		if l == strings.ToUpper(p.Letter) {
			return &p
		}
	}
	return nil
}

// ParameterOrDefault will return the Parameter for the given letter or return the given default value.
// Lookup is case-insensitive.
func (c *Code) ParameterOrDefault(letter string, value interface{}) *CodeParameter {
	p := c.Parameter(letter)
	if p != nil {
		return p
	}
	return NewSimpleCodeParameter(letter, value)
}

// ReplaceParameter will replace the first occurrence of a parameter with the given letter
func (c *Code) ReplaceParameter(letter string, np *CodeParameter) bool {
	for i, p := range c.Parameters {
		if p.Letter == letter {
			c.Parameters[i] = *np
			return true
		}
	}
	return false
}

// RemoveParameter removes all parameters with the given letter
func (c *Code) RemoveParameter(letter string) *CodeParameter {
	p := c.Parameter(letter)
	if p != nil {
		cp := make([]CodeParameter, 0)
		for _, p := range c.Parameters {
			if p.Letter != letter {
				cp = append(cp, p)
			}
		}
		c.Parameters = cp
	}
	return p
}

// GetUnprecedentedString reconstructs an unprecedented string from parameter list
func (c *Code) GetUnprecedentedString(quote bool) string {
	var b strings.Builder
	for _, p := range c.Parameters {
		if b.Len() > 0 {
			b.WriteString(" ")
		}
		b.WriteString(p.Letter)
		if quote && p.IsString {
			b.WriteString(`"`)
		}
		b.WriteString(p.AsString())
		if quote && p.IsString {
			b.WriteString(`"`)
		}
	}
	return b.String()
}

// String will convert the parsed code back to a text-based G/M/T-code
func (c *Code) String() string {
	if c.Keyword != None {
		if c.KeywordArgument != "" {
			return fmt.Sprintf("%s %s", c.Keyword.String(), c.KeywordArgument)
		}
		return c.Keyword.String()
	}
	if c.Type == Comment {
		return ";" + c.Comment
	}
	var b strings.Builder
	b.WriteString(c.ShortString())

	for _, p := range c.Parameters {
		b.WriteString(" ")
		b.WriteString(p.String())
	}

	if c.Comment != "" {
		if b.Len() > 0 {
			b.WriteString(" ")
		}
		b.WriteString(";")
		b.WriteString(c.Comment)
	}

	if len(c.Result) > 0 {
		b.WriteString(" => ")
		b.WriteString(strings.TrimRight(c.Result.String(), " "))
	}

	return b.String()
}

// ShortString converts only the command portion to text-based G/M/T-code (e.g. G28)
func (c *Code) ShortString() string {
	if c.Keyword != None {
		return c.Keyword.String()
	}
	if c.Type == Comment {
		return "(comment)"
	}

	var prefix string
	if c.HasFlag(EnforceAbsolutePosition) {
		prefix = "G53 "
	}
	if c.MajorNumber != nil {
		if c.MinorNumber != nil {
			return fmt.Sprintf("%s%s%d.%d", prefix, c.Type, *c.MajorNumber, *c.MinorNumber)
		}
		return fmt.Sprintf("%s%s%d", prefix, c.Type, *c.MajorNumber)
	}
	return fmt.Sprintf("%s%s", prefix, string(c.Type))
}
